### PR TITLE
Add error for non-`Proc`-typed block argument

### LINKED ISSUE #6483: Sorbet should not allow non-Proc types for `&blk` arguments
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20CallTakesBlock%0A%20%20extend%20T%3A%3ASig%0A%20%20sig%20%7Bparams%28x%3A%20Integer%29.void%7D%0A%20%20def%20call%28x%29%0A%20%20%20%20p%20x%0A%20%20end%0A%0A%20%20sig%20%7Breturns%28CallTakesBlock%29%7D%0A%20%20def%20to_proc%0A%20%20%20%20self%0A%20%20end%0Aend%0A%0Asig%20%7Bparams%28blk%3A%20CallTakesBlock%29.void%7D%0Adef%20example%28%26blk%29%0A%20%20yield%2042%0A%20%20yield%20''%0Aend%0A%0A%23%20runtime%20error%3A%20%60CallTakesBlock%23to_proc%60%20returns%20neither%20nil%20nor%20a%20%60Proc%60%20object%0A%23%20https%3A%2F%2Fgithub.com%2Fruby%2Fruby%2Fblob%2Fv2_7_6%2Fvm_args.c%23L1150-L1154%0Aexample%28%26CallTakesBlock.new%29)

```ruby
# typed: true
extend T::Sig

class CallTakesBlock
  extend T::Sig
  sig {params(x: Integer).void}
  def call(x)
    p x
  end

  sig {returns(CallTakesBlock)}
  def to_proc
    self
  end
end

sig {params(blk: CallTakesBlock).void}
def example(&blk)
  yield 42
  yield ''
end

# runtime error: `CallTakesBlock#to_proc` returns neither nil nor a `Proc` object
# https://github.com/ruby/ruby/blob/v2_7_6/vm_args.c#L1150-L1154
example(&CallTakesBlock.new)
```

#### Observed output

```
editor.rb:20: Expected Integer but found String("") for argument x https://srb.help/7002
    20 |  yield ''
                ^^
  Expected Integer for argument x of method CallTakesBlock#call:
    editor.rb:6:
     6 |  sig {params(x: Integer).void}
                      ^
  Got String("") originating from:
    editor.rb:20:
    20 |  yield ''
                ^^
Errors: 1
```

#### Expected behavior

The Ruby VM requires that `to_proc` [must convert to `nil` or a `Proc`](https://github.com/ruby/ruby/blob/v2_7_6/vm_args.c#L1150-L1154). That means that the `blk` argument can never bind to a non-`Proc` object, which means that Sorbet should reject all signatures that type the `&blk` argument as something other than a (possibly nilable) `Proc` (note that internally, Sorbet models `T.proc.void` and friends as generic subclasses of the `Proc` class, so doing a simple `isSubType(..., Proc)` should be sufficient.

- - -

### PR DESCRIPTION
**Pull request by jez**
### Motivation

Fixes #6483

The Ruby VM does not allow this.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_MdfW1NQlJLbJsV
→ https://go/builds/bui_MdfWyr8Bkurxqp

**Comment by jez** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6484** <a href="https://app.graphite.dev/github/pr/sorbet/sorbet/6484" target="_blank">Graphite</a> 
    * **PR #6486** <a href="https://app.graphite.dev/github/pr/sorbet/sorbet/6486" target="_blank">Graphite</a>  👈
      * **PR #6487** <a href="https://app.graphite.dev/github/pr/sorbet/sorbet/6487" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sorbet/sorbet/6486?utm_source=stack-comment).

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_MdgHIDMkIMepod
→ https://go/builds/bui_MdgIJZO7YDXSbO

File: resolver/resolver.cc [hunk 1]
```diff
@@ -3435,18 +3435,30 @@ class ResolveSignaturesWalk {
...
                 arg.type = std::move(spec->type);
-                arg.loc = spec->loc;
+                if (isBlkArg && !core::Types::isSubType(ctx, arg.type, core::Types::nilableProcClass())) {
+                    if (auto e = ctx.state.beginError(spec->nameLoc, core::errors::Resolver::InvalidMethodSignature)) {
+                        e.setHeader("Block argument type must be either `{}` or a `{}` type (and possibly nilable)",
+                                    "Proc", "T.proc");
+                        e.addErrorNote(
+                            "Ruby allows classes to have custom `{}` methods to convert `{}`-passed arguments into\n"
+                            "    `{}` objects, but even those methods must return `{}` objects.",
+                            "to_proc", "&blk", "Proc", "Proc");
```
Comment (froydnj) [inline resolver/resolver.cc:3455]: OOC, do we have tests for honoring this magic `to_proc` behavior?

Comment (jez) [reply resolver/resolver.cc:3455]: sort of but they could be better. i added one.

**Pull request merged**